Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp

```diff
@@ -69,60 +70,82 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
+    };
+}
 
-    HashMap<WebCore::FrameIdentifier, Vector<WebFoundTextRange>> frameMatches;
-    for (auto& simpleRange : findMatches) {
-        Ref document = simpleRange.startContainer().document();
+static inline bool canConvertToWebFoundTextRange(SimpleRange& range)
+{
+    Ref document = range.startContainer().document();
 
-        RefPtr element = document->documentElement();
-        if (!element)
-            continue;
+    RefPtr element = document->documentElement();
+    if (!element)
+        return false;
 
-        RefPtr frame = document->frame();
-        if (!frame)
-            continue;
+    RefPtr frame = document->frame();
+    if (!frame)
+        return false;
 
-        // FIXME: We should get the character ranges at the same time as the SimpleRanges to avoid additional traversals.
-        auto range = characterRange(makeBoundaryPointBeforeNodeContents(*element), simpleRange, WebCore::findIteratorOptions());
-        auto domData = WebFoundTextRange::DOMData { range.location, range.length };
-        // order set by UI process
-        auto foundTextRange = WebFoundTextRange { domData, frame->pathToFrame(), 0 };
-        m_cachedFoundRanges.add(foundTextRange, simpleRange.makeWeakSimpleRange());
+    return true;
+}
 
-        const auto frameID = frame->frameID();
-        auto& matches = frameMatches.ensure(frameID, [] {
-            return Vector<WebFoundTextRange> { };
-        }).iterator->value;
+#if ENABLE(PDF_PLUGIN)
+static inline Vector<WebFoundTextRange::PDFData> findPdfMatchesInFrame(Frame* frame, const String& string, const OptionSet<FindOptions>& options)
+{
+    RefPtr localFrame = dynamicDowncast<LocalFrame>(frame);
+    if (!localFrame)
+        return { };
+
+    RefPtr pluginView = WebPage::pluginViewForFrame(localFrame.get());
+    if (!pluginView)
+        return { };
 
+    return pluginView->findTextMatches(string, core(options));
+}
+#endif
+
+void WebFoundTextRangeController::findTextRangesForStringMatches(const String& string, OptionSet<FindOptions> options, uint32_t maxMatchCount, CompletionHandler<void(HashMap<WebCore::FrameIdentifier, Vector<WebFoundTextRange>>&&)>&& completionHandler)
+{
+    auto matchingRanges = protectedWebPage()->protectedCorePage()->findTextMatches(string, core(options), maxMatchCount, false);
+    Vector<WebCore::SimpleRange> findMatches = WTFMove(matchingRanges.ranges);
+
+    if (findMatches.size() > 0)
+        m_cachedFoundRanges.clear();
+
+    auto validSimpleRanges = findMatches | std::views::filter(canConvertToWebFoundTextRange);
+    auto webFoundTextRanges = validSimpleRanges | std::views::transform(createWebFoundTextRange);
+
+    const auto createEmptyVector = []() {
+        return Vector<WebFoundTextRange> { };
+    };
+
+    HashMap<WebCore::FrameIdentifier, Vector<WebFoundTextRange>> frameMatches;
+    for (const auto& [foundTextRange, simpleRange] : std::views::zip(webFoundTextRanges, validSimpleRanges)) {
```

Comment:
This is also another iteration over matches. But I guess the real bottleneck is the `FIXME` about creating the `CharacterRange`.